docs/general/builtins/scalar-functions.rst

```diff
@@ -2995,6 +2995,42 @@ Returns: ``array(text)``
     SELECT 1 row in set (... sec)
 
 
+.. _scalar-object_merge:
+
+``object_merge(object, object)``
+--------------------------------
+
+The ``object_merge`` function combines two objects into a new object 
+containing the union of their first level properties, taking the second 
+object's values for duplicate properties.
+
```

Comment:
Maybe worth to add a more prominent note about unsupported recursive merge?
E.g.
```
.. NOTE:

  Does not operate recursively: only the top-level array or object structure is merged.
```

And maybe even show this behaviour by an example?

```
    cr> select
    ...     object_merge({a = 1, b = { d = 4 }}, {a = 2, b = { c = 2 }}) AS object_merge;
    +-------------------------+
    | object_merge            |
    +-------------------------+
    | {"a": 2, "b": {"c": 2}} |
    +-------------------------+
    SELECT 1 row in set (... sec)
```

Reply:
From a quick check on other functions, maybe it's worth mentnioning also, what happens with `null`s .
if both are null, null is returned, if one of the 2 is null, the non-null is returned.

Reply:
Is there a reason we don't want to support recursive merge?

Reply:
> Is there a reason we don't want to support recursive merge?

I think if supported, then it should be optional.
The PostgreSQL JSONB-concat also only operates on the top-level.

Reply:
I'd also support this optional to be compliant with PG (at least from the operator aspect) and probably for performance reasons. 
Couldn't it be an optional 3rd parameter, either a flag to enable nested merge or a level integer?
But could be also done as a followup.

Reply:
I'd propose to only expose the `||` syntax for now (and use a hidden function name prefixed with `_`) that mirrors PostgreSQL semantics.

We can then follow up with a more powerful variant that lets users also configure the behavior, e.g. there could be something like `keep`, `error`, `force` - a bit inspired from neovim APIS:

```
tbl_deep_extend({behavior}, {...})                     *vim.tbl_deep_extend()*
    Merges recursively two or more map-like tables.

    Parameters:  
      • {behavior}  (string) Decides what to do if a key is found in more than
                    one map:
                    • "error": raise an error
                    • "keep": use value from the leftmost map
                    • "force": use value from the rightmost map
      • {...}       (table) Two or more map-like tables

    Return:  
        (table) Merged table
```

Reply:
> I'd propose to only expose the || syntax for now (and use a hidden function name prefixed with _) that mirrors PostgreSQL semantics.

should this then be exposed in the documentation as `concat(object, object)`, which `||` is an alias for?
In PostgreSQL `concat` is seemingly always doing string-concatenation, while in CrateDB it also concatenates `ARRAYS`

If you think `object_merge` might imply a deep merge, `object_assign` (following JS nomenclature https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign?retiredLocale=de). would be an alternative?

Reply:
Would need some input here to finish this up

- use `object_merge` (or another dedicated name) and `concat` as alias (for pg compatibility)
  -  if a deep merge should be supported in the future this could be handled through additional arguments. There is no "standard" way to do this - that I am aware - anyway.
- use `concat` (with the alias `||`) 
  - how shall this be documented (As extension to the existing `concat` or as `concat` entry in the object functions?)
- use `||` only
  - this would imply a change in the parser / ast, as `left || right` is currently an alias for `concat(left,right)`. Also doesn't really fit with other scalar functions.

Reply:
We currently map all `||`  to `concat` as you said and adding a differentiation based on the left/right type isn't really possible at the parser level so we would have to add additional logic later on to rewrite it to a different (internal) function. @mfussenegger Or do you have another idea in mind how to implement `||` only without much complexity?

Otherwise, in order to not add complexity, I'd prefer:

> - use `object_merge` (or another dedicated name) and `concat` as alias (for pg compatibility)
> 
>     * if a deep merge should be supported in the future this could be handled through additional arguments. There is no "standard" way to do this - that I am aware - anyway.

But instead of adding and documenting an explicit alias to `concat` this could be an document alias to `||`. That it can be then, due to our `||` implementation, accessed via `concat` as well, I wouldn't document.

Reply:
added a `object_deepmerge(object, object)` variant as well (in 2nd commit).
Don't really know if this needs a "behaviour" left or right precedence can be achieve by swapping inputs.

Reply:
> added a `object_deepmerge(object, object)` variant as well (in 2nd commit).
> Don't really know if this needs a "behaviour" left or right precedence can be achieve by swapping inputs

I agree on the swapping part but what is about the "raise error on duplicates" part? I think this is still interesting.

Reply:
> I agree on the swapping part but what is about the "raise error on duplicates" part? I think this is still interesting.

when would an exception be raised for a deep merge?
For any duplicate key, or only if a non-object value would be overwritten? Imho it would only make sense for the later, or otherwise one should not use a deep merge.

Reply:
Why wouldn't it make sense to raise an error if e.g. a nested object value is overwritten by a primitive one?

Reply:
Can you make a separate PR for the second commit? Discussion on the interface/behavior regarding the deep_merge doesn't need to block the simple merge function

Reply:
@mfussenegger @seut 

I moved the deep merge to a separate PR (https://github.com/crate/crate/pull/13370) 

---

Could you please advise how to finalise the simple merge

Reply:
I'd go with 2) use concat (with the alias ||) 

As for docs: maybe both, mention in the concat docs that object type is supported and explain the behavior and in the object functions section add a link/reference

Reply:
adjusted towards 2) in latest commit